File: api/api/routers/payments.py
import os

import stripe
from fastapi import APIRouter
from pydantic import BaseModel, Field, model_validator

from api.dependencies.security import RequiredUserDep, RequiredUserOrganizationDep
from api.dependencies.services import PaymentServiceDep, PaymentSystemServiceDep
from api.services.payments_service import PaymentMethodResponse, PaymentService
from api.tags import RouteTags
from core.utils.background import add_background_task

router = APIRouter(prefix="/organization/payments", tags=[RouteTags.PAYMENTS])


stripe.api_key = os.getenv("STRIPE_API_KEY")


class PaymentMethodRequest(BaseModel):
    payment_method_id: str
    payment_method_currency: str = "USD"


class PaymentMethodIdResponse(BaseModel):
    payment_method_id: str


@router.post("/payment-methods", description="Add a payment method to the organization")
async def add_payment_method(
    payment_service: PaymentServiceDep,
    request: PaymentMethodRequest,
    user_org: RequiredUserOrganizationDep,
) -> PaymentMethodIdResponse:
    return PaymentMethodIdResponse(
        payment_method_id=await payment_service.add_payment_method(user_org, request.payment_method_id),
    )


class CustomerCreatedResponse(BaseModel):
    customer_id: str


@router.post("/customers", description="Create a customer in Stripe for the organization")
async def create_customer(
    payment_service: PaymentServiceDep,
    userDep: RequiredUserDep,
) -> CustomerCreatedResponse:
    return CustomerCreatedResponse(customer_id=await payment_service.create_customer(userDep.sub))


class PaymentIntentCreatedResponse(BaseModel):
    client_secret: str = Field(
        description="The stripe client secret for the payment intent, "
        "that can be used by the client to retrieve the payment using the client secret and a publishable key.",
    )
    payment_intent_id: str


class CreatePaymentIntentRequest(BaseModel):
    amount: float


@router.post("/payment-intents", description="Create a manual payment intent in Stripe for the organization")
async def create_payment_intent(
    request: CreatePaymentIntentRequest,
    user_org: RequiredUserOrganizationDep,
) -> PaymentIntentCreatedResponse:
    payment_intent = await PaymentService.create_payment_intent(user_org, request.amount, trigger="manual")
    if not payment_intent.client_secret:
        raise ValueError("Payment intent creation failed")
    return PaymentIntentCreatedResponse(
        client_secret=payment_intent.client_secret,
        payment_intent_id=payment_intent.payment_intent_id,
    )


@router.get("/payment-methods", description="Get the payment method attached to the organization")
async def get_payment_method(
    user_org: RequiredUserOrganizationDep,
) -> PaymentMethodResponse | None:
    return await PaymentService.get_payment_method(user_org)


@router.delete("/payment-methods", description="Delete the payment method attached to the organization")
async def delete_payment_method(
    payment_service: PaymentServiceDep,
    user_org: RequiredUserOrganizationDep,
) -> None:
    await payment_service.delete_payment_method(user_org)


class AutomaticPaymentRequest(BaseModel):
    opt_in: bool
    threshold: float | None = None
    balance_to_maintain: float | None = None

    @model_validator(mode="after")
    def validate_threshold_and_balance_to_maintain(self: "AutomaticPaymentRequest") -> "AutomaticPaymentRequest":
        if self.opt_in:
            if self.threshold is None or self.balance_to_maintain is None:
                raise ValueError("Threshold and balance_to_maintain are required when opt_in is true")
            if self.threshold > self.balance_to_maintain:
                raise ValueError("Threshold must be greater than balance_to_maintain")
        else:
            if self.threshold is not None or self.balance_to_maintain is not None:
                raise ValueError("Threshold and balance_to_maintain must be None when opt_in is false")
        return self


@router.put("/automatic-payments", description="Enable or disable automatic payments")
async def update_automatic_payments(
    request: AutomaticPaymentRequest,
    payment_service: PaymentServiceDep,
    payment_system_service: PaymentSystemServiceDep,
    user_org: RequiredUserOrganizationDep,
) -> None:
    await payment_service.configure_automatic_payment(
        opt_in=request.opt_in,
        threshold=request.threshold,
        balance_to_maintain=request.balance_to_maintain,
    )
    if request.opt_in:
        # Trigger a credit decrement if 9 to trigger a payment if needed
        add_background_task(payment_system_service.decrement_credits(user_org.tenant, 0))


@router.post("/automatic-payments/retry", description="Retry a failed automatic payment")
async def retry_automatic_payment(
    payment_service: PaymentSystemServiceDep,
    user_org: RequiredUserOrganizationDep,
) -> None:
    await payment_service.retry_automatic_payment(user_org)

File: client/src/store/payments.ts
import { produce } from 'immer';
import { useEffect } from 'react';
import { create } from 'zustand';
import { client } from '@/lib/api';
import {
  AutomaticPaymentRequest,
  CreatePaymentIntentRequest,
  CustomerCreatedResponse,
  PaymentIntentCreatedResponse,
  PaymentMethodResponse,
} from '@/types/workflowAI';
import { useOrganizationSettings } from './organization_settings';

interface PaymentsState {
  isLoading: boolean;

  isPaymentMethodInitialized: boolean;

  paymentMethod: PaymentMethodResponse | undefined;

  addPaymentMethod: (paymentMethodId: string) => Promise<void>;
  getPaymentMethod: () => Promise<void>;
  createPaymentIntent: (amount: number) => Promise<PaymentIntentCreatedResponse>;
  updateAutomaticPayment: (optIn: boolean, threshold: number | null, balanceToMaintain: number | null) => Promise<void>;
  deletePaymentMethod: () => Promise<void>;
}

export const usePayments = create<PaymentsState>((set) => ({
  isLoading: false,

  isPaymentMethodInitialized: false,

  paymentMethodId: undefined,
  paymentMethod: undefined,

  getPaymentMethod: async () => {
    try {
      const response = await client.get<PaymentMethodResponse | null>(
        `/api/data/organization/payments/payment-methods`
      );
      set(
        produce((state: PaymentsState) => {
          state.paymentMethod = !!response ? response : undefined;
        })
      );
    } catch (error) {
      console.error(error);
    } finally {
      set(
        produce((state: PaymentsState) => {
          state.isPaymentMethodInitialized = true;
        })
      );
    }
  },

  addPaymentMethod: async (paymentMethodId: string) => {
    set(
      produce((state: PaymentsState) => {
        state.isLoading = true;
      })
    );
    try {
      await client.post<{ payment_method_id: string }, PaymentMethodResponse>(
        `/api/data/organization/payments/payment-methods`,
        {
          payment_method_id: paymentMethodId,
        }
      );
    } catch (error) {
      throw error;
    } finally {
      set(
        produce((state: PaymentsState) => {
          state.isLoading = false;
        })
      );
    }

    await usePayments.getState().getPaymentMethod();
  },

  createPaymentIntent: async (amount: number) => {
    const response = await client.post<CreatePaymentIntentRequest, PaymentIntentCreatedResponse>(
      `/api/data/organization/payments/payment-intents`,
      {
        amount,
      }
    );

    return response;
  },

  updateAutomaticPayment: async (optIn: boolean, threshold: number | null, balanceToMaintain: number | null) => {
    try {
      await client.put<AutomaticPaymentRequest>(`/api/data/organization/payments/automatic-payments`, {
        opt_in: optIn,
        threshold,
        balance_to_maintain: balanceToMaintain,
      });
      await useOrganizationSettings.getState().fetchOrganizationSettings();
    } catch (error) {
      console.error(error);
    }
  },

  deletePaymentMethod: async () => {
    try {
      await client.del<void>(`/api/data/organization/payments/payment-methods`);
      await useOrganizationSettings.getState().fetchOrganizationSettings();
      await usePayments.getState().getPaymentMethod();
    } catch (error) {
      console.error(error);
    }
  },

  retryAutomaticPayment: async () => {
    await client.post(`/api/data/organization/payments/automatic-payments/retry`, undefined);
    await useOrganizationSettings.getState().fetchOrganizationSettings();
  },
}));

export const useOrFetchPayments = () => {
  const paymentMethod = usePayments((state) => state.paymentMethod);
  const isLoading = usePayments((state) => state.isLoading);
  const isInitialized = usePayments((state) => state.isPaymentMethodInitialized);
  const getPaymentMethod = usePayments((state) => state.getPaymentMethod);

  useEffect(() => {
    if (!isInitialized) {
      getPaymentMethod();
    }
  }, [getPaymentMethod, isInitialized]);

  return {
    paymentMethod,
    isLoading,
    isInitialized,
  };
};
